File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

N, M = map(int, sys.stdin.readline().split())
trees = list(map(int, sys.stdin.readline().split()))

start, end = 0, max(trees)

while start <= end :
    mid = (start + end) // 2
    result = 0

    for i in trees :
        if i >= mid :
            result += i - mid
    if result >= M :
        start = mid + 1
    else :
        end = mid - 1

print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		long M = Integer.parseInt(st.nextToken());

		long min = Integer.MAX_VALUE;
		long max = 0;
		
		long[] trees = new long[N];
		
		st = new StringTokenizer(br.readLine());
		
		for (int i = 0; i < N; i++) {
			trees[i] = Integer.parseInt(st.nextToken());
			min = 0;
			max = Math.max(max, trees[i]);
		}
		
		while (min <= max) {
			long mid = (min + max) / 2;
			long check = M;
			
			for (long i : trees) {
				if (i > mid) {
					check -= (i - mid);
				}
			}
			
			if (check > 0) {
				max = mid - 1;
			} else {
				min = mid + 1;
			}
		}
		
		System.out.println(max);
		br.close();
	}
}